### PR TITLE
feat: Add unique secondary index

### LINKED ISSUE #298: Sec. Indexes: Implement unique indexes
Part of https://github.com/sourcenetwork/defradb/issues/201

Use composite keys

### PR DESCRIPTION
**Pull request by islamaliev**
## Relevant issue(s)

Resolves #298 

## Description

Introduces unique secondary indexes.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

Unit and integration tests

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: db/fetcher/indexer_iterators.go
Comment (fredcarle) [inline db/fetcher/indexer_iterators.go:63]: suggestion: I would leave the error as a seperate return value. This makes it more obvious to the caller that the error needs to be checked.

File: db/fetcher/indexer_iterators.go
Comment (fredcarle) [inline db/fetcher/indexer_iterators.go:102]: suggestion: Same as above. It would be nicer if ther error was kept separate.

File: db/collection_index.go
Comment (fredcarle) [inline db/collection_index.go:226]: question: Since this operation is happening on a transaction, why do we have to bother removing the indexed documents if it fails? Can we not rely on the discarding of the transaction?

File: db/fetcher/indexer_iterators.go
Comment (islamaliev) [reply db/fetcher/indexer_iterators.go:102]: I followed what seems to be a common way of dealing with iterators. Like `query.Result`.
I think the main advantage of it (besides having shorter returns) it will be easier to use such API with go channels.

File: db/collection_index.go
Comment (islamaliev) [reply db/collection_index.go:226]: I thought about it myself, but wasn't sure if it would be correct to discard the whole transaction. After all, it theoretically might contain several operations (not only index creation) it should be up to the caller's site to decide what to do.
On the other hand, calling `colIndex.RemoveAll` is very localized and the collection has all necessary knowledge to make this decision. 
Maybe I'm missing something.
Note: here is the test that tests it https://github.com/sourcenetwork/defradb/pull/2131/files#diff-0a00fc4743373305afbef72902d95cdc5a9ddbda32c694a0c97c85d03efcde46R1034

File: db/fetcher/indexer_iterators.go
Comment (fredcarle) [reply db/fetcher/indexer_iterators.go:102]: Do you think we will move to go channels for that API? I'm not sure a channel makes sense here.

**Review by fredcarle — APPROVED** [review]
Overall it look good. I'm approving assuming we will resolve the 2 conversations before merging.

**Pull request merged**
